src/musekafka/message.py

```diff
@@ -0,0 +1,192 @@
+        )
+        batch_duration += datetime.utcnow() - batch_start
+        if not batch:
+            # Empty batch does not count towards num_batches,
+            # since we require batch_size to be > 0.
+            continue
+        LOGGER.debug("Got batch of %d messages.", len(batch))
+        num_batches += 1
+        yield batch
+
+    LOGGER.debug(
+        "Completed streaming %d batches, with each batch of size at most %d.",
+        num_batches,
+        batch_size,
+    )
+
+
+def stream(
+    consumer: Consumer,
+    count: Optional[int] = None,
+    poll_timeout: Optional[float] = DEFAULT_POLL_TIMEOUT,
+    timeout: Optional[float] = None,
+) -> Iterator[Message]:
+    """Stream `count` messages from the consumer."""
+    if count is not None and count <= 0:
+        raise ValueError("count must be a positive integer.")
+    if timeout is not None and timeout <= 0:
+        raise ValueError("timeout must be a positive float.")
+
+    if count is None:
+        LOGGER.debug("Streaming messages....")
+    else:
+        LOGGER.debug("Streaming up to %d messages....", count)
+
+    if timeout is not None:
+        timeout_delta = timedelta(milliseconds=int(timeout * MILLIS_IN_SECOND))
+
+    stream_duration = timedelta()
+    num_messages = 0
+    while count is None or num_messages < count:
+        if musekafka.shutdown.is_shutting_down():
+            break
+
+        if timeout is not None:
+            if stream_duration >= timeout_delta:
+                LOGGER.debug("Hit stream timeout (%.3f seconds).", timeout)
+                break
+            poll_timeout = min(
+                poll_timeout or timeout, (timeout_delta - stream_duration).total_seconds()
+            )
+
+        stream_start = datetime.utcnow()
+        try:
+            message = poll(consumer, timeout=poll_timeout)
+        except TimeoutError as e:
+            LOGGER.debug(str(e))
+            continue
+        finally:
+            stream_duration += datetime.utcnow() - stream_start
+        ctx = dict(
+            kafka_topic=message.topic(),
+            kafka_partition=message.partition(),
+            kafka_offset=message.offset(),
+        )
+        with _bind_ctx(ctx):
+            yield message
+        num_messages += 1
+    LOGGER.debug("Completed streaming %d messages.", num_messages)
+
+
+def poll(consumer: Consumer, timeout: Optional[float] = DEFAULT_POLL_TIMEOUT) -> Message:
+    """Poll for a message from the consumer.
+
+    This method differs from consumer.poll in that it will never return a null message
+    or a message that has an error. If you set timeout and poll hits that timeout,
+    it will raise a TimeoutError.
+    """
+    if timeout is not None and timeout <= 0:
+        raise ValueError("timeout must be a positive integer.")
+
+    if timeout is not None:
+        timeout_delta = timedelta(milliseconds=int(timeout * MILLIS_IN_SECOND))
+
+    poll_duration = timedelta()
+    while True:
```

Comment:
If this is polling a single message why is this in a `while True`?

Reply:
Oh I think I see, it polls until it gets a non null message right?

Reply:
Yes. Same logic that is in the consumer in our primary repository.